### PR TITLE
pam: also set teleport-specific env vars via pam_putenv

### PR DESCRIPTION
**Pull request by awly**
Using `pam_putenv` from `libpam` exposes these env vars to `pam_exec.so`
and possibly other built-in PAM modules. Keep setting them via
`os.Setenv` too, for `pam_script.so` to use.

Updates #3692

### CODE REVIEW

File: lib/pam/pam.go [hunk 1]
```diff
@@ -280,6 +283,20 @@ func Open(config *Config) (*PAM, error) {
...
 	}
 
+	for k, v := range config.Env {
+		// Set a regular OS env var on this process which should be available
+		// to child PAM processes.
+		os.Setenv(k, v)
+
+		// Also set it via PAM-specific pam_putenv, which is respected by
+		// pam_exec (and possibly others), where parent env vars are not.
+		kv := C.CString(fmt.Sprintf("%s=%s", k, v))
```
Comment (russjones) [inline lib/pam/pam.go:293]: `C.CString` does a malloc under the hood. Will PAM free this memory? If not you'll need to call `C.free` yourself.

File: lib/pam/pam.go [hunk 2]
```diff
@@ -280,6 +283,20 @@ func Open(config *Config) (*PAM, error) {
...
+		retval := C._pam_putenv(pamHandle, p.pamh, kv)
```
Comment (russjones) [inline lib/pam/pam.go:294]: Put `pam_putenv` have any maximum size? If so you might want to cap the length here.

File: lib/pam/pam.go (hunk 1 above)
Comment (awly) [reply lib/pam/pam.go:293]: Ah, nice catch, added `C.free`.
First time using cgo, appreciate all the advice!

File: lib/pam/pam.go (hunk 2 above)
Comment (awly) [reply lib/pam/pam.go:296]: http://www.linux-pam.org/Linux-PAM-html/adg-interface-by-app-expected.html#adg-pam_putenv doesn't mention any limits and their source code seems to allocate as much as needed.

**Review by fspmarshall — APPROVED** [review]

**Review by russjones — APPROVED** [review]

**Pull request merged**